### PR TITLE
ERC: Secret Multisig Recovery

### PR DESCRIPTION
**Pull request by 3esmit**
Read latest version at https://gitlab.com/status-im/docs/EIPs/blob/secret-multisig-recovery/EIPS/eip-2429.md

### CODE REVIEW

**Comment by vbuterin** [comment]
This feels like it could be simplified more. Why not this version:

* We save into a contract a Merkle root `recovery_root = merkle_root([address1, threshold, secret1], [address2, threshold, secret2] ...)`
* We give every recovery participant the sister leaves to provide a Merkle proof of their own address, the threshold and secret (BTW we also need to standardize the encoding of this, so that you can eg. paste the data to someone via chat application if they're not using the same app as you)
* To recover, you call their contract and provide the Merkle proof, along with some `recoverTo` data and an `authorizationWitness`. If your address is an EOA, then the `authorizationWitness` must be an ethereum signed message. If your address is a contract, then the recovery should call a `checkAuthorized(recoverTo, authorizationWitness) -> bool` method of your address.

All other details, eg. how the secret per user get computed, can be left to individual wallets to figure out.

**Comment by vbuterin** [comment]
BTW it seems worthwhile to make `checkAuthorized` a separate ERC, generalizing the notion of ethereum signed messages to contract accounts. It could be usable in many other cases. Not sure if this has been done already.

**Comment by 3esmit** [comment]
Hey @vbuterin , thanks for the interest in this EIP and for the suggestions, I know you care a lot about social recovery and I am happy to see you here. 

One goal of Secret Multisig Recovery is that, the `secret_set` holder don't have to trust the "friends" helping to recover the account. In the current version: 
1. the secret-set holder can store the secret-set in not so safe places (we can call it a semi-secret), like at a drawer (with a tamper evident seal) or for the careless in a cloud storage service, because it only reveals how to start a recovery, and wallets should suggest a manual review of request, perhaps with video call;  
2. the addresses selected from the merkle tree just know they are part of recovery at an event; 
3. the secret holder can prove they know the secret-set by showing `hash_to_peer` and `merkle proof`, and wallets must check this before notifying someone to help recover; 
4. peers have to sign an obfuscated `calldata` parameter, that basically is `keccak256(address(this),hash_to_execute,to,data)`, which only gets revealed at execution, making overruns impossible in normal gas market conditions, which wallets can advice to don't operate recovery when used gas in blocks are near block gas limit. 
5. `hash_to_execute` will be unique for every recovery

To start a recovery an attacker must have the secret-set and would have to social engineering  people from recovery list (or have compromised a very large set of accounts), which all are very difficult to set, therefore drawing down interest in attacking Secret Multisig Recovery. 
The social engineering could be done with help of AI deep fakes rendering specifically "the account owner persona" (image and voice), however for building a neural network of a person it needs many images of that person (the more the better) and voice samples. This already looks like 007 level of hacking, and so I considered it safe enough.

Seems like your main suggestion is to move the secret to the merkle leafs, so they don't have to be stored or checked at execute, and maybe execute can be removed (executes automatically when threshold is met)? That is an interesting suggestion, and would simplify the contract.

I will investigate further your suggestion to see if possible to keep the "complete trustless" property, because from a brief view seems like it don't supports the above point `4.`, and this would open a (very small) window for collusion, where in a (very specific) scenario of "bad friends" (or an attacker with access to many accounts) could overrun the recovery transaction with their own parameters (that could even split funds back to these addresses using a contract). 
Do you think this risk is unrealistic or maybe we can mitigate collusion in a different way, perhaps by allowing only one approval per leaf? 

Regarding `checkAuthorized`, this EIP uses [EIP-1271](https://eips.ethereum.org/EIPS/eip-1271): `isValidSignature(bytes _data, bytes _signature) -> bytes4` for that, but if another better standard emerges I am happy to change, as I don't like too much that it returns `bytes4 MAGIC_VALUE` stuff that tries to solve problems only happening in old solidity or badly programmed contracts, however community consensus and conventions first.

Account contracts not implementing [EIP-1271](https://eips.ethereum.org/EIPS/eip-1271) can still help a recovery by making a direct call to `approve`, or they can be externally owned accounts. see this [snippet taken from body of  `approvePreSigned`](https://github.com/status-im/account-contracts/blob/ec61340390bd9721856003a675116947e28821a0/contracts/account/MultisigRecovery.sol#L172-L176):
```solidity
require(
    (
        isContract(_signer) && ERC1271(_signer).isValidSignature(abi.encodePacked(signingHash), _signature) == EIP1271_MAGICVALUE
    ) || ECDSA.recover(signingHash, _signature) == _signer,
    "Invalid signature");
```
The message (`_data`) to be signed is constructed with [EIP-191](https://eips.ethereum.org/EIPS/eip-191)v0 standard, but I am planning to move to [EIP-712](https://eips.ethereum.org/EIPS/eip-712) when [EIP-1344](https://eips.ethereum.org/EIPS/eip-1344) gets incorporated to it.
I still have much to improve in this standard, such as allowing infinite thresholds, which currently is limited by block gas limit because of a for loop in execute, but its possible to count it in the approval (and leave junk on state, forever).

**Comment by 3esmit** [comment]
> All other details, eg. how the secret per user get computed, can be left to individual wallets to figure out.

Based on your suggestion, I added the `secret_type` parameter to the recovery private URL, and added 3 types to start with. Also added `paramters`, as I found it could be useful for setting parameters on different secret_types. This would help wallets figure out and define their own standards.

**Comment by vbuterin** [comment]
Ah, I see. So the recovery data is not even provided to the friends until the recovery is warranted. I didn't even realize that this was a desired objective. IMO it's not clear that this is a good idea, because it limits some use cases of recovery, particularly where the account holder is incapacitated or dies. Sure, you could just give the recovery data to someone else, but then if you do that then why not just make that person be one of the recovery participants?

Basically, the challenge is that you're IMO needlessly increasing complexity because you're splitting one functionality (authorization) across two different mechanisms: knowledge of the secret, and control over the address. You can simplify greatly by just putting all the authorization into control over addresses, and then use the secrets mechanism only to solve the problem of ensuring privacy of who the guardians for some address are (without secrets, an attacker with medium computation power can just loop over all combinations of addresses until they find a matching Merkle root).

For example, suppose that you want a scheme with the following property. An account can be recovered if 3 of 5 of (Alice, Bob, Charlie, David, Evan) approve the recovery, but only after a secret S is revealed and the holder of secret S approves the request (S could be stored by the account holder or in a drawer). Here is how you would replicate this within my framework.

Make an **8 of 10** recovery, between (Alice, Bob, Charlie, David, Evan, S_1, S_2, S_3, S_4, S_5), where `S_i` is an address derived from a private key derived from S. That is: `S_i = privtoaddr(hash(S, i))`. Then, to start a recovery, you would need to generate the 5 S_i keys and generate signatures for them, and then ask your friends to recover. The friends by themselves cannot do anything without S.

Glad that ERC 1271 exists! I somehow did not know about it :D That definitely solves that problem.

**Comment by 3esmit** [comment]
> this was a desired objective

Actually, the desired objective is not having to trust any address in the list at any point.

> it limits some use cases of recovery, particularly where the account holder is incapacitated or dies

I am also interested in having a "dead man switch" in account contracts, but in this case I believe that it must require some sort of account inactivity, which would enable another account (publicly known) to take over it, and I think this case should be handled at the key holder side with a watchdog looking for calls from the owner. 

Anyway, if the secret-set is available, is very likely that a close relative would be able to solve the challenge (biometrics, user form) and maybe convince people in the list to recover the address.  



> For example, suppose that you want a scheme with the following property. An account can be recovered if 3 of 5 of (Alice, Bob, Charlie, David, Evan) approve the recovery, but only after a secret S is revealed and the holder of secret S approves the request (S could be stored by the account holder or in a drawer). Here is how you would replicate this within my framework.
> 
> Make an **8 of 10** recovery, between (Alice, Bob, Charlie, David, Evan, S_1, S_2, S_3, S_4, S_5), where `S_i` is an address derived from a private key derived from S. That is: `S_i = privtoaddr(hash(S, i))`. Then, to start a recovery, you would need to generate the 5 S_i keys and generate signatures for them, and then ask your friends to recover. The friends by themselves cannot do anything without S.

Basically instead of using secret reveal, it uses EC with the secret as a private key.  This does not reduces any UX complexity, however it is indeed a more simple solution in the smart contract side.  
To simplify it could be that the owner defines a validator address which is controlled by the private key from user secret data, being this the only key that could decide the recovery call.

Do you think that presenting the merkle proof to the peers is enough to validate requests (just to prevent spam in this channel, still need human verification)?

**Comment by 3esmit** [comment]
I am still considering your suggestion @vbuterin, as it using less concepts is better, and I see that it would be safer to use it as EC, as you never in fact reveal any private piece. 

However, the secret is not used only for: 
- preventing tx overrun;
- prevent collusion;

But also to: 
- reuse the merkle tree of friends (thats why peer hash is used inside merkle leafs),
- to obfuscate the threshold for until the execution is possible.

So as the public hash must change, and its created based on recovery address + nonce, the merkle root, threshold and public hash will always be unique, and will be very hard for use past recovery revealed data to reconstruct the new merkle tree, as the new peer_hash is missing, and even if one attacker knows all the addresses from merkle tree, they still wont be able to verify it

I see that maybe I could hash the public address and use the non hashed address to obfuscate the merkle tree, or maybe I can keep using the secret and use at some point derive an special address that reveals. 

I will have to think more on how to implement this without using commit-reveal, if I don't find anywhere to fit this here, I will be happy in creating another EIP simillar to this to specify this other version. and also make sure that both `Secret Recovery URL` and `Recovery Request URL` support versioning for the recovery systems.

**Comment by vbuterin** [comment]
> reuse the merkle tree of friends (thats why peer hash is used inside merkle leafs),

Can you elaborate on what you mean by this?

> to obfuscate the threshold for until the execution is possible.

You mean obfuscate it from the world, or obfuscate from the participants themselves?

**Comment by 3esmit** [comment]
> > reuse the merkle tree of friends (thats why peer hash is used inside merkle leafs),
> 
> Can you elaborate on what you mean by this?

When the merkle tree is revealed, it becomes public (or partially public), because you need to proof it to the contract. We expect that after a recovery, people keeps using recovery, but we don't want to loose any of the security assumptions, such as the most basic the merkle tree of multisig privacy. 

So, the standard is requesting that every recovery setup derives an unique merkle tree, because of the User secret data hashing, that's why there is a nonce in the recovery that is used privately in the hashing to generate the revealed secret (`Execute Hash`). The contract itself prevents that the same `Public Hash` is used, to  avoid the possibility of a known merkle tree be used.

> > to obfuscate the threshold for until the execution is possible.
> 
> You mean obfuscate it from the world, or obfuscate from the participants themselves?

I mean to obfuscate from everyone, until the reveal of `Execute Hash`. To enhance the security, Threshold value is kept private. This was done because the "user private secret" is needed anyway, and so we try to hide everything as possible. 
This is also interesting because someone with few "friends" could use a small threshold without compromising any security (assuming that the secret-set is kept private). Without knowing the threshold, an attacker would not be able to detect weaker recoveries, meaning that all would look to be equally difficult to break until the secret-set is known. 

BTW, I see that this system could possibly enable trustless automated recovery services with KYC, only based on the "Peer Hash" verification, even if only a few trustful signers are needed.
To enabled that I think that I we should add a "weight" value to the leafs (revealed at approval), so someone could use a single "KYC recovery service" or many of their "friends", trustless, as secret set is needed to recover.

**Comment by vbuterin** [comment]
I feel like revealing the threshold when the first recovery leaf is submitted is okay; the time period between that and the final recovery leaf is going to be very small relative to the entire time that the account lives.

> To enabled that I think that we should add a "weight" value to the leafs

So this actually opens up another approach: you can set the threshold to be a fixed 100 points, and use weights to determine the actual threshold. This way you could support differing weights, and you even get a bit of uncertainty around the threshold, because you could do a 3-of-5 using weights like eg. 50 33 20 48 48.

**Comment by vbuterin** [comment]
> When the merkle tree is revealed, it becomes public (or partially public), because you need to proof it to the contract. We expect that after a recovery, people keeps using recovery, but we don't want to loose any of the security assumptions, such as the most basic the merkle tree of multisig privacy. So, the standard is requesting that every recovery setup derives an unique merkle tree,

Agree that this privacy protection is important! Though it is achieved just by having _any_ kind of secret in the leaves, and changing it when you re-do the setup, no?

**Comment by ngmachado** [comment]
> I feel like revealing the threshold when the first recovery leaf is submitted is okay; the time period between that and the final recovery leaf is going to be very small relative to the entire time that the account lives.
> 
> > To enabled that I think that we should add a "weight" value to the leafs
> 
> So this actually opens up another approach: you can set the threshold to be a fixed 100 points, and use weights to determine the actual threshold. This way you could support differing weights, and you even get a bit of uncertainty around the threshold, because you could do a 3-of-5 using weights like eg. 50 33 20 48 48.

Adding weight for each leaf, also enables Inheritance plan like a 'dead man switch'. This contract will contain some logic but the interface will probability be define in another ERC. This one is getting particular big.

Im still worry about the privacy of the peers and the account link after a reveal. This privacy setting will also probably be better define in another ERC.

**Comment by 3esmit** [comment]
> So this actually opens up another approach: you can set the threshold to be a fixed 100 points, and use weights to determine the actual threshold. This way you could support differing weights, and you even get a bit of uncertainty around the threshold, because you could do a 3-of-5 using weights like eg. 50 33 20 48 48.

Having a fixed threshold is indeed a good idea, however then we change the `weight` to be mandatory (as they usually won't be 1). As the weight was added, it makes sense to keep it on the merkle tree leaves only.

> Agree that this privacy protection is important! Though it is achieved just by having _any_ kind of secret in the leaves, and changing it when you re-do the setup, no?

Yes, any kind of secret. So, currently user have in the setup phase to give `public hash`, that at recovery works like an normal commit-reveal scheme, however it uses two hashing stages to enable a "partial reveal" to make easier to validate the recovery request without having to give the revealing hash, that is also what is used to commit with the secret call. 
Is only one secret that user have to use, that is constructed from the user secret set.

> > I feel like revealing the threshold when the first recovery leaf is submitted is okay; the time period between that and the final recovery leaf is going to be very small relative to the entire time that the account lives.
> > > To enabled that I think that we should add a "weight" value to the leafs
> > 
> > 
> > So this actually opens up another approach: you can set the threshold to be a fixed 100 points, and use weights to determine the actual threshold. This way you could support differing weights, and you even get a bit of uncertainty around the threshold, because you could do a 3-of-5 using weights like eg. 50 33 20 48 48.
> 
> Adding weight for each leaf, also enables Inheritance plan like a 'dead man switch'. This contract will contain some logic but the interface will probability be define in another ERC. This one is getting particular big.
> 
> Im still worry about the privacy of the peers and the account link after a reveal. This privacy setting will also probably be better define in another ERC.

I think that inheritance plan must be a dedicated contract, and public, and have another type of rules, such a "watchdog contract" for inactivity, where if the owner dont pings for every period (year, configurable), another address (or a list of addresses, or whatever rules of inheritance) can claim which would take a delay time. When a claim is started, the original owner would still be able to ping to refuse the claim, in case they missed. 

There are other questions surrounding dead man switch and account contracts, in case of inheritance, the account contract could not be used anymore (because it's a "dead identity") but it would still enable some sort of actions, such as withdraw of ERC20/ERC721 tokens.  This is important because we want to be nice with reputation systems that uses nontransferable tokens into account contracts. However this is arguable. We can start the discussion about this issue in a separated EIP or forum, and I would be happy to contribute to it as well, however I want to first focus in solutions that I have already researched for a while now.  

I think we should keep all "Secret Multisig Recovery" into one ERC, additional EIPs could be used to extend this one, but I agree that we should plan for extensibility, as this is a good property for network effects.

**Comment by 3esmit** [comment]
I defined the fixed threshold to be `100 * 10^18`, following the standard for divisibility in ether and used `100` units to make easier for influence % calculations on UI.

**Comment by vbuterin** [comment]
Is what's on the github above right now intended to be the latest version?

**Comment by 3esmit** [comment]
Yes, I try to keep the PR in sync with latest version of file committed to this PR, as well on this link because of `mermaid` graphs support https://hackmd.io/@3esmit/rJ4YBV_JI 

Regarding the "Implementations", I am updating right now the solidity implementation on [Status.im Account-Contracts Repository](https://github.com/status-im/account-contracts/blob/develop/contracts/account/MultisigRecovery.sol) to fit the latest updates on ABI and the fixed threshold, but I didnt implemented test cases yet because of the constant changes.

**Comment by 3esmit** [comment]
> If so, my main remaining question is, why not make the "structure" behind `hash_to_peer` optional and out of scope?

The hashing schema is there to make sure `hash_to_peer` is unique, specifically by salting the `private hash` with `recovery_address + recovery_address.nonce()`, but this all happens in private, to create the `hash_to_execute`. 
The reason of having `hash_to_execute` is to allow `secret_call` being kept secret even with `peer_to_hash` revealed.
We could make it optional and the wallet use some "default" fields (like `hash_to_execute` being `bytes32(0)`), but this would worsen a lot the security properties, as allowing collusion and (partially) undisclosed merkle tree after recovery. 

The EIP suggest that a "web2 model a-like" of "security phrase" or "biometrics", however this is not enough as this "semi-private" data can be guessed, therefore we also need the approval of "friends" to make really difficult an attack. As the merkle tree is already a "secret", the additional secret helps keeping this merkle tree secret.  

> I foresee that the lazy thing many people will want to do is not bother with the structure, and just put `hash_to_peer` into the help request. 

The EIP-831 based URLs in this EIP have a lot of room for improvement. For now the helprecover URL should contains all the needed parameters to help signing the approval, but some, like merkle proof is optional, only required if the friend have to call directly, but I think that there is no problem in passing the merkle proof. 

The process will not be manually handled by users, instead a UI should enforce users to use the standard correctly.

> Also, why not allow the option of putting `hash_to_peer` into the help request?

Ah, good point, I named incorrectly `peer_hash` there, sorry about that, just fixed it.

**Comment by 3esmit** [comment]
Perhaps you mean putting out of scope the generation of "user_secret_data"? I thought in pre-defining some initial standards, such as biometrics, password derived and user data questionary, as options, but keeping it open to other standards. 

Thinking here that ethereum have in production gnosis safe with recovery using regular (transparent) multisigs, and we might want to make at specify the helprequest URL for those.

Perhaps we also need to improve the EIP in regards of future proofing the URLs, or defining the EIP-831 URLs in a new EIP (or one for each) that would handle all the multiple recovery types we can have. 
So only the individual parameters used in the URL that this EIP requires would be defined here. 

For now I would be keeping all in one place while trying to keep future proof, because it's easier to make all in one EIP, and when someone make it later (maybe myself) we can start spinning off parts of this EIPs.

**Comment by 3esmit** [comment]
I'll be added other members in the discussion as authors to help ensure this is kept alive:
- Miguel Mota (@miguelmota); 
- Vitalik Buterin (@vbuterin);
- naxe (@ngmachado).

If you disagree, just emoji thumbs down here that I remove your name. I will also be adding other participants as they contribute to the discussion.

**Comment by 3esmit** [comment]
I started addeing some mock up of UIs to help visualize the concept. I decided using https://www.plantuml.com/ because it's markdownish syntax is understandable even without the SVG render. See how cool it looks on VSCodium:
image

**Comment by 3esmit** [comment]
Btw, in devcon 4 (prague) I did this mock ups at a UX stuff, sorry for the horrible drawing, I am better with code.
IMG_20181101_124316

**Comment by ngmachado** [comment]
Starting a new Topics:
https://ethereum-magicians.org/t/membership-proof-in-a-merkle-tree-with-peer-secret/3904

Just to focus on the privacy and relationship between peers and user of recovery contract

**Comment by vbuterin** [comment]
Should we just make a telegram chat or some other more synchronous medium to go through all the details? Feels like there's a lot of moving parts here and it would be good to quickly get on the same page.

**Comment by 3esmit** [comment]
Sure, I am available in telegram. My username in telegram is @ricagsch (telegram don't allow usernames starting with a number)

**Comment by 3esmit** [comment]
Ah, I sent this EIP .md to gitlab (because gitlab is better) https://gitlab.com/status-im/docs/EIPs/blob/secret-multisig-recovery/EIPS/eip-2429.md 

It supports the fancy stuff im using, as plantuml and mermaid, and its open source.

**Comment by 3esmit** [comment]
Sure, here is the invitation link for telegram group https://t.me/joinchat/HPycUBTRBNM2lS1Bz5Ytzw

**Comment by vbuterin** [comment]
Here's my proposal for reworking the design to achieve a few more properties:

* One single recovery contract
* It's not known which address is recovering until recovery time
* Different hash_to_peer for each user, providing more privacy

First, the "secrets" hash structure:

```
hash(private_hash) = execute_hash
hash(execute_hash) = public_hash
hash(execute_hash + peer_address) = hash_to_peer
```

This way, each leaf has a different `hash_to_peer` so revealing one leaf won't allow brute force attacks that can unmask other leaves, as those other leaves will still have their own still-unknown `hash_to_peer` values.

Now, the hash structure for commitments:

```
         [commitment_hash]
        /                 \
[merkle_root]        [execute_hash]
      /  \
    ...  /\    
       ... \
          [address, weight, hash_to_peer, is_ens]
```

There is one single recovery contract, R. Suppose that a wallet W wants to set up a recovery.

#### SETUP PROCEDURE

1. The owner of `W` determines the (address, weight, is_ens) values and computes a `private_hash`
2. `W` computes `commitment_hash`. `W` calls `R`'s method `register`, passing `commitment_hash` as an argument. `R` saves `self.commitments[W] = commitment_hash` in its storage

#### RECOVERY PROCEDURE

1. Each participant publishes their merkle branches (going up to `merkle_root` only) in a call to `R`'s `approve` method. The contract counts the total weight supporting each `(merkle_root, secret_call)` pair, and stores a bitfield to prevent double calls from one address
2. When the weight supporting some pair is sufficient, the owner of `W` (technically, anyone can do this, but in practice it would be the owner of `W`) can call `R`'s `execute` method, revealing `execute_hash`, `calldata` and `destination` (= `W`). `R` checks that `h(execute_hash, calldata, W)` actually equals `secret_call` and checks that `self.commitments[W] = h(execute_hash, merkle_root)`. If both checks pass, `R` calls `W` with the given `calldata`.

**Comment by 3esmit** [comment]
Thanks a lot @vbuterin for the contributions. I created this PR https://github.com/status-im/account-contracts/pull/1 that updates a implementation listed in the EIP with your insight.

**Comment by 3esmit** [comment]
Also, now the spec is updated as mentioned above, but might be missing update on some old parts of old spec. I'll keep reviewing it.

**Comment by vbuterin** [comment]
In private chats I proposed an even simpler version; I'll copy it here as well for more public discussion.

### Hash tree structure

```python
leaves = [(address, weight, is_ens, hash_to_peer1), (address, weight, is_ens, hash_to_peer2) ...]
leaf_hashes = [hash(leaves[0]), hash(leaves[1]) ...]
leaf_root = hash(leaf_hashes)
commitment_hash = hash(leaf_root, execute_hash)
```

#### SETUP PROCEDURE (Unchanged from above)

* The owner of `W` determines the `(address, weight, is_ens)` values and computes a `private_hash`
* The owner of `W` computes `commitment_hash`. `W` calls `R`'s method register, passing `commitment_hash` as an argument. `R` saves `self.commitments[W] = commitment_hash` in its storage

#### RECOVERY PROCEDURE

* Each participant submits the full contents of their leaf along with `secret_call` (hash of `calldata` + `destination` + `execute_hash`) and a signature to `R`'s `submit_leaf` method. Upon verifying the signature, `R` saves `self.approved_weights[hash(hash(leaf), secret_call)] = leaf.weight`
* When the weight supporting some execute hash for some address is sufficient, the owner of `W` (technically, anyone can do this, but in practice it would be the owner of `W`) can call `R`'s execute method, revealing `execute_hash`, `calldata` and `destination` (= W) along with a list of all `leaf_hash` values. `R` computes `commitment_hash = hash(hash(leaf_roots), execute_hash)` and checks that `self.commitments[W] = commitment_hash`. It also computes `secret_call`, computes `hash(leaf_hash + secret_call)` for every `leaf_hash`, accesses those positions in `self.approved_weights` and verifies that the sum of the returned values is sufficient. If both checks pass, R clears those values out of storage and calls W with the given calldata.

This removes the need for Merkle trees, decreases total costs in many cases, and creates a higher degree of privacy because it is not known on-chain whether or not two recovery approvals even refer to the same wallet, or which wallet it is, before the finalization step.

#### Optimizations

As written, the gas cost for the `execute` step is 512 for each `leaf_hash` + 800 for SLOAD + ~100 for SHA3 and other logistics, so ~1400 total, plus 5000 for storage clearing (which is refunded). If desired, we could require `execute` to provide a bitfield of which leaves are nonzero, reducing the number of SLOAD operations required by restricting them to nonzero leaves; this would reduce per-user gas costs by more than half and hence facilitate eg. 10-of-200 recoveries. This would also be good forward-planning for a future where SLOAD becomes even more expensive for stateless client reasons.

We could optimize further by using a [multiproof](https://github.com/ethereum/eth2.0-specs/blob/dev/ssz/merkle-proofs.md#merkle-multiproofs) instead of requiring every leaf, but this greatly increases complexity and is IMO unlikely to be worth it.

**Comment by 0zAND1z** [comment]
Hi, I came across this thread on twitter and I see a lot of progress.

Can someone help me contrast this to Shamir Secret Sharing?

**Comment by johnsBeharry** [comment]
@0zAND1z With Shamir secret sharing — you would split your private key into shares which you seek out contacts/locations to distribute them to yourself.

The entire ceremony of recovery is a loose process which involves you essentially hunting down the shares and piecing them together to get the original secret back.

#### PROBLEMS
0. SSS does not allow for weighted shares, each share is made equal so you would need to split a key with multiple thresholds
1. Contacts: You need to maintain a list of these contacts/locations for recovery.
2. Share Storage: These contacts need to have secure locations to store the shares. Since a share cannot be revoked/made invalid if a share is compromised, depending on the threshold security is either lost completely or by some large factor.
3. Recovery Tool: There are no standard implementation for SSS so a successful recovery is also tool dependent (what if the software is no longer maintained).

---

*This is a recovery system for a Smart Contract Wallet, not simply just Private Key/Secret recovery.*

@3esmit proposal aims to replicate "forgot password" in a "trust limiting" manner thats analogous to the web2 apps we use now but depending on multiple "guardians" to allow the "password reset".

*e.g. Currently an app would allow you to overwrite the password they have for your account that's stored in their DB if you're able to provide a secret "mothers maiden name".*

Question is how do we replicate this in a trustless blockchain enviroment where there isn't a central person able to recover your account and you're the one with full ownership / access. AFAIK there isn't any customer service 1800-ETHEREUM numnber to call to do this recovery (@vitalk ?) so we distribute this trust to > 1 parties who can help us with recovery.

You set a secret — e.g. "mothers maden name" as with the web2 example above.

You go to a subset of the a group of addresses `address_list` you had defined on setup (`secret_set`) to grant a new address you own access to the Smart Contract Wallet. *They never own a fragment of the origonal secret as with SSS.*

These addresses sign a message which you gather and call a recover method on the contract — they do not need to find new ways to secure a SSS share instead, they would just secure their own private keys as normal.

> *only secret holder can specify the recovery call parameters that `address_list` can authorize.*

This creates a easy to standardise — distrubuted recovery system that can be applied to any smart contract.

1. No shares to distribute
2. Set weights for each address, not all are equal (5 shares a: 10%, a: 10%, a: 20%, a: 20%, a: 40%)
3. Address list + Recovery Secret you must still manage and store securely

@3esmit did I get this right? Sorry for the length — hopefully it was useful.

**Comment by ngmachado** [comment]
> @0zAND1z With Shamir secret sharing — you would split your private key into shares which you seek out contacts/locations to distribute them to yourself.
> 
> The entire ceremony of recovery is a loose process which involves you essentially hunting down the shares and piecing them together to get the original secret back.
> 
> #### PROBLEMS WITH SSS & SOCIAL RECOVERY
> 1. SSS does not allow for weighted shares, each share is made equal so you would need to split a key with multiple thresholds
> 2. Contacts: You need to maintain a list of these contacts for recovery.
> 3. Share Storage: These contacts need to have secure locations to store the shares. Since a share cannot be revoked/made invalid if a share is compromised, depending on the threshold security is either lost completely or by some large factor.
> 4. Recovery Tool: There are no standard implementation for SSS so a successful recovery is also tool dependent (what if the software is no longer maintained / you need to know which curve was used).
> 
> _This is a recovery system for a Smart Contract Wallet, not simply just Private Key/Secret recovery._
> 
> @3esmit proposal aims to replicate "forgot password" in a "trust limiting" manner thats analogous to the web2 apps we use now but depending on multiple "guardians" to allow the "password reset".
> 
> _e.g. Currently an app would allow you to overwrite the password they have for your account that's stored in their DB if you're able to provide a secret "mothers maiden name"._
> 
> Question is how do we replicate this in a trustless blockchain enviroment where there isn't a central person able to recover your account and you're the one with full ownership / access. AFAIK there isn't any customer service 1800-ETHEREUM numnber to call to do this recovery (@vitalk ?) so we distribute this trust to > 1 parties who can help us with recovery.
> 
> You set a secret — e.g. "mothers maden name" as with the web2 example above.
> 
> You go to a subset of the a group of addresses `address_list` you had defined on setup (`secret_set`) to grant a new address you own access to the Smart Contract Wallet. _They never own a fragment of the origonal secret as with SSS._
> 
> These addresses sign a message which you gather and call a recover method on the contract — they do not need to find new ways to secure a SSS share instead, they would just secure their own private keys as normal.
> 
> > _only secret holder can specify the recovery call parameters that `address_list` can authorize._
> 
> This creates a easy to standardise — distrubuted recovery system that can be applied to any smart contract.
> 
> 1. No shares to distribute
> 2. Set weights for each address, not all are equal (2/3 of 5 shares a: 10%, b: 10%, c: 20%, d: 20%, e: 40%)
> 3. Address list + Recovery Secret you must still manage and store securely
> 4. All Etheruem addresses become this contact list esp. when paired with an ENS domain allows for contact details lookups.
> 5. Recovery Software lives as a smart contract
> 
> @3esmit did I get this right? Sorry for the length — hopefully it was useful.

Also try to protect the users from one hand of models attacks, so if the user take in consideration the racional of this ERC we will get a better privacy, even for the Peers (Guardians)

Just to clarify one point, the recovery process can execute an arbitrary call (destination and data). Setting a new Owner is only one use case.

**Comment by 3esmit** [comment]
> @3esmit did I get this right? Sorry for the length — hopefully it was useful.

@johnsBeharry  Thanks for the explanation, you are absolutely correct on that points. 

I would also add that: 
- Only owner of wallet contract knows who are the Guardians, the guardian are only asked for help if this is ever needed;
- It's possible to set the entire earth population as guardians  (see sparse merkle trees) and it costs the same (see merkle root and SSZ multi proofs);
- No one knows if you have only one guardian or the entire population of earth as guardians;
- Guardians don't know how much guardians are needed to recover;
- Guardians can have different weights for approving the recovery;
- The compromise of "user secret" don't results in loss, only in the leakage of the guardian list.
- It's a simple method, besides using advanced concepts like sparse merkle trees and multi proofs, everything is based on a smart commit-reveal hashing of the user secret + approval of guardians, which can be formally proven.

File: EIPS/eip-2429.md [hunk 1]
```diff
@@ -0,0 +1,830 @@
...
+
+
+## Simple Summary
+
+Allows user to select a private list of ethereum addresses which can authorize a call protected by a secret only user knows. 
+Allows account contracts to set a private list of ethereum addresses that can authorize a call protected by a secret.
+
+## Abstract
+
+A bad experience of crypto-currencies in general is the lost or exposure of a private key, which can lead to irreversible situations.
```
Comment (d10r) [inline EIPS/eip-2429.md]: I guess _the lost_ should be replaced with _the loss_ in this sentence

Comment (3esmit) [reply EIPS/eip-2429.md]: Thanks a lot for the review! There are probably other "English errors" in the text, specially in cohesion and correct derivation of words. I'm not a native English speaker and after the EIP is Final, not only a security audit is needed, but also an English audit :smile: 

I'll this open until a commit fixes it.

**Comment by shanefontaine** [comment]
Has there been discussion about the possibilities of ENS names expiring or being traded in the case where `is_ens = true`?

If a user sets their ENS name as the recoverer and that ENS name expires (or they sell it), a malicious actor could buy the name and recover the account, correct? Is there any protection against this except for messaging to the user that their name is tied to this recovery (as opposed to the address).

Does the address vs ENS name become more of a social issue (as opposed to technical) with this EIP? Some thoughts:

* Some people will think that their Ethereum address is their canonical source of truth, while others may think the opposite.
* People could compare ENS to DNS, at which point they would assume their ENS name is the source of truth  and the Ethereum address (IP Address in DNS) is abstracted away. This may be amplified by applications that intentionally try to only show the  ENS name in the name of better UX.

**Comment by ngmachado** [comment]
> If a user sets their ENS name as the recoveree and that ENS name expires (or they sell it), a malicious actor could buy the name and recover the account, correct? Is there any protection against this except for messaging to the user that their name is tied to this recovery (as opposed to the address)?

In general a good practice to select a number of addresses or ENS as a recovery option greater than the threshold. By register a sufficient number of peers, that risk can be minimised. 

Example : Selecting two peers is not strong enough to a long standing solution.

I want to make clear one point, registering peers is not equal to use all that peers. You can attribute a height to each peer so you can later make a good selection based on you social realm and work from that point.

About sending messages to the Peer, i don't like that solution, the peers should only be notify in a recovery process to protect the user needing a recovery against others type of attacks. 

Please see the [Security Considerations](https://gitlab.com/status-im/docs/EIPs/blob/secret-multisig-recovery/EIPS/eip-2429.md#security-considerations) to get more datails.

Using the ENS is a good addition to this ERC but each user is not required to use it. By not using the ENS you still maintain functionality but lose one important feature.

**Comment by 3esmit** [comment]
> Has there been discussion about the possibilities of ENS names expiring or being traded in the case where `is_ens = true`?

As @ngmachado already said, the only risk here is not being possible to approve the recovery. That's because the list of guardians is part of the "user secret set", which never becomes public. 
Guardians are only revealed after the recovery claiming their signature, but after the recovery is complete, the "user secret hash" is updated trough a nonce and hash process, altering the whole merkle tree, rendering any information about guardians require also knowing the user secret and the adjacent merkle leafs of the selected guardians (otherwise the merkle proofs cannot be created). 
The revealed guardians are not the revealed leafs, but they are the composition of the leafs of a merkle tree that changes at every nonce, even if the guardian list remains the same, and is impossible to tell if the guardian list changed.

**Comment by 3esmit** [comment]
Still active work being done, although the active discussion is external.
I'll review it again and prepare a working PoC for it to be ready to be merged.

**Comment by MicahZoltu** [comment]
I'll wait for your upcoming changes to do a formal review, but as a teaser I'm going to say something along the lines of:

This sounds like a Good Idea, not a good Standard.  Something can be a Good Idea, or even a Great Idea, but not need a standard.  Pretty much all design patterns fall into this category, and pretty much every application that is popular/good falls into this category.  All are good ideas, none of them need standards.

As an example, I have written https://github.com/Zoltu/recoverable-wallet/ which I *personally* think is a Good Idea, but I don't think it needs a standard.  I think everyone should be using Recoverable Wallets, but there isn't really a benefit to everyone using this exact implementation of a recoverable wallet.
